trunk/agarrcsd/agarrcsd.8
.\"
.\" Copyright (c) 2007 Hypertriton, Inc. <http://hypertriton.com/>
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 
.\" THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
.\" IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
.\" WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT,
.\" INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
.\" (INCLUDING BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
.\" SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
.\" STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING
.\" IN ANY WAY OUT OF THE USE OF THIS SOFTWARE EVEN IF ADVISED OF THE
.\" POSSIBILITY OF SUCH DAMAGE.
.\"
.Dd SEPTEMBER 4, 2007
.Dt AGARRCSD 8
.Os
.ds vT Agar API Reference
.ds oS Agar 1.3
.Sh NAME
.Nm agarrcsd
.Nd agar-rcs server
.Sh SYNOPSIS
.Nm agarrcsd
.Op Fl av
.Op Fl b Ar hostname
.Op Fl p Ar port
.Op Fl d Ar datadir
.Sh DESCRIPTION
The
.Nm
daemon implements a very simple, specialized revision control system for
Agar objects (see
.Xr AG_Object 3 ) .
When compiled with network support, Agar-Core provides a built-in client
interface to
.Nm .
.Pp
The options are as follows:
.Bl -tag -width Ds
.It Fl b
Bind the server to the given address (hostname or a numerical address).
.It Fl p
Bind the server to the given port.
.It Fl a
Start the administrator interface instead of the daemon.
.El
.Sh SERVER COMMANDS
.Pp
.Nm
implements only a few simple commands:
.Bl -tag -width "rcs-commit "
.It version
Return the server version string.
.It rcs-commit
Commit a new revision to a given object.
If the object does not exist on the repository, it is created.
Parameters are as follows:
.Pp
.Bl -tag -compact -width "object-digest "
.It object-path
Absolute object pathname in repository.
.It object-size
Size of dataset in bytes.
.It object-name
Object name (relative to parent object).
.It object-type
Object class name.
.It object-digest
Checksums for object dataset.
.It commit-msg
Commit message.
.El
.Pp
If the parameters are accepted, the server reponds with a successful status
code of 0, and then reads
.Fa object-size
bytes of data.
If the transfer is successful, the server returns a successful status code
with a message of the form
.Dq New revision: #<number> .
.It rcs-info
Return revision information for the given object.
Parameters are as follows:
.Pp
.Bl -tag -compact -width "object-path "
.It object-path
Absolute pathname to the object.
.It revision
Optional revision number.
If this parameter is not specified, information about the latest revision
is returned.
.El
.Pp
The server will then return a colon-separated string of the form:
.Bd -literal
r=<rev-number>
a=<author>
d=<digest>
m=<commit-msg>
t=<object-class>
n=<object-name>
.Ed
.It rcs-update
Download the dataset of the specified object.
Parameters are as follows:
.Pp
.Bl -tag -compact -width "object-path "
.It object-path
Absolute pathname to the object.
.It object-name
Object name relative to parent.
.It object-type
Object class string.
.It revision
Revision number to download.
.El
.Pp
If successful, the server will return the dataset in binary format.
.It rcs-list
Return the list of all repository entries.
Individual list entries are of the form:
.Bd -literal
<path>:<rev>:<author>:<obj-class>:<obj-name>
.Ed
.It rcs-log
Return a list of all revisions for the given object.
Parameters are as follows:
.Pp
.Bl -tag -compact -width "object-path "
.It object-path
Absolute path to object in repository.
.El
.Pp
The command will return a list of entries of the form:
.Bd -literal
<rev>:<author>:<type>:<name>:<digest>:<commit-msg>
.Ed
.El
.\" .Sh ENVIRONMENT
.\" .Sh FILES
.Sh SEE ALSO
.Xr AG_Intro 3 ,
.Xr AG_Object 3 ,
.Xr NS 3 ,
.Xr NC 3


trunk/agarrcsd/agarrcsd.c
/*
 * Copyright (c) 2007 Hypertriton, Inc. <http://hypertriton.com/>
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE FOR
 * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
 * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
 * USE OF THIS SOFTWARE EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

#include <config/version.h>

#include <sys/types.h>
#include <sys/time.h>
#include <sys/socket.h>
#include <sys/wait.h>

#include <netinet/in.h>
#include <netinet/in_systm.h>
#include <netinet/ip.h>

#include <stdlib.h>
#include <stdio.h>
#include <stdarg.h>
#include <string.h>
#include <unistd.h>
#include <netdb.h>
#include <errno.h>
#include <fcntl.h>
#include <signal.h>

#include "agarrcsd.h"
#include "pathnames.h"
#include "protocol.h"
#include "rcs.h"

#include <agar/gui.h>
#include <agar/dev.h>

volatile int gothup = 0;
User *user;
NS_Server *serv;

static void
sig_hup(int sigraised)
{
	gothup++;
}

static int
version(NS_Server *serv, NS_Command *cmd, void *p)
{
	char hostname[128];

	gethostname(hostname, sizeof(hostname));
	NS_Message(serv, 0, "agarrcsd:%s:%s", VERSION, hostname);
	return (0);
}

static int
auth_password(NS_Server *serv, void *p)
{
	char buf[65], *pBuf = &buf[0];
	char *name, *pass;
	User *u;
	size_t end;

	if (fgets(buf, sizeof(buf), stdin) == NULL)
		return (-1);

	if ((name = strsep(&pBuf, ":")) == NULL ||
	    (pass = strsep(&pBuf, ":")) == NULL) {
		return (-1);
	}
	end = strlen(pass)-1;
	if (pass[end] == '\n')
		pass[end] = '\0';

	if ((u = UserLookup(name)) == NULL) {
		NS_Log(NS_INFO, "%s: no such user", name);
		goto fail;
	}
	if (strcmp(pass, u->pass) == 0) {
		user = u;
		NS_Log(NS_INFO, "%s: login successful", name);
		return (1);
	}
	NS_Log(NS_INFO, "%s: password mismatch", name);
fail:
	AG_SetError("Unknown username or password mismatch");
	return (0);
}

static int
auth_pubkey(NS_Server *serv, void *p)
{
	AG_SetError("Public key auth is not implemented");
	return (0);
}

int
main(int argc, char *argv[])
{
	extern const AG_ObjectOps UserOps;
	char *host = NULL;
	char *port = "6785";
	char *dir = _PATH_DATA;
	User *u;
	struct sigaction sa;
	int adminflag = 0;
	int i;
	
	if (AG_InitCore("agarrcsd", 0) == -1) {
		fprintf(stderr, "%s\n", AG_GetError());
		return (1);
	}
	while ((i = getopt(argc, argv, "ab:p:d:v?")) != -1) {
		extern char *__progname;

		switch (i) {
		case 'a':
			adminflag = 1;
			break;
		case 'b':
			host = optarg;
			break;
		case 'p':
			port = optarg;
			break;
		case 'd':
			dir = optarg;
			break;
		case 'v':
			printf("%s %s\n", __progname, VERSION);
			exit(0);
		case '?':
		default:
			printf("Usage: %s [-av] [-b host] [-p port] [-d dir]\n",
			    __progname);
			exit(0);
		}
	}

	AG_RegisterType(&UserOps, OBJ_ICON);

	if (adminflag) {
		/* Set up administrator interface. */
		if (AG_InitVideo(640, 480, 32, AG_VIDEO_RESIZABLE) == -1 ||
		    AG_InitInput(0) == -1) {
			fprintf(stderr, "%s\n", AG_GetError());
			return (-1);
		}
		AG_InitConfigWin(0);
		AG_SetRefreshRate(-1);
		AG_BindGlobalKey(SDLK_ESCAPE, KMOD_NONE, AG_Quit);
		AG_BindGlobalKey(SDLK_F1, KMOD_NONE, AG_ShowSettings);

		DEV_InitSubsystem(0);
		DEV_Browser();
	}

	/* Reload the previous state (including user data). */
	AG_ObjectLoad(agWorld);
	AGOBJECT_FOREACH_CLASS(u, agWorld, user, "User:*")
		AG_ObjectLoad(u);

	/* Move to the data directory. */
	if (chdir(dir) == -1) {
		fprintf(stderr, "%s: %s\n", dir, strerror(errno));
		exit(1);
	}

	/* Initialize the server. */
	serv = NS_ServerNew(NULL, 0, "_server", _PROTO_NAME, _PROTO_VER, port);
	NS_ServerBind(serv, host, port);

	/* Set up the SIGHUP handler. */
	sa.sa_flags = 0;
	sa.sa_handler = sig_hup;
	sigaction(SIGHUP, &sa, NULL);

	/* Register our auth methods and error functions. */
	NS_RegAuthMode(serv, "password", auth_password, NULL);
	NS_RegAuthMode(serv, "public-key", auth_pubkey, NULL);

	/* Register the daemon functions. */
	NS_RegCmd(serv, "version", version, NULL);
	NS_RegCmd(serv, "rcs-commit", rcs_commit, NULL);
	NS_RegCmd(serv, "rcs-info", rcs_info, NULL);
	NS_RegCmd(serv, "rcs-update", rcs_update, NULL);
	NS_RegCmd(serv, "rcs-list", rcs_list, NULL);
	NS_RegCmd(serv, "rcs-log", rcs_log, NULL);

	if (adminflag) {
		AG_EventLoop();
		goto out;
	}

	/* Start server main loop. */
	if (NS_Listen(serv) == -1) {
		fprintf(stderr, "%s.%s: %s\n", host, port, strerror(errno));
		exit(1);
	}
out:
	AG_Destroy();
	return (0);
}
